From 4d3f36ccf432bca230736ef88e44c4478ed4cfd1 Mon Sep 17 00:00:00 2001
From: Katsuya Matsubara <matsu@igel.co.jp>
Date: Fri, 1 Jun 2012 15:41:07 +0900
Subject: [PATCH 32/80] rcarvin: support 12/18/24 bit bus width

VIN could support not only 8 and 16 bit but also 12, 18, and
24 bit bus width.

Signed-off-by: Katsuya Matsubara <matsu@igel.co.jp>
---
 drivers/media/video/rcarvin.c |   17 +++++++++++++++--
 include/media/rcarvin.h       |    9 ++++++---
 2 files changed, 21 insertions(+), 5 deletions(-)

diff --git a/drivers/media/video/rcarvin.c b/drivers/media/video/rcarvin.c
index 10e974a..b39c337 100644
--- a/drivers/media/video/rcarvin.c
+++ b/drivers/media/video/rcarvin.c
@@ -209,9 +209,18 @@ static unsigned long make_bus_param(struct rcar_vin_dev *pcdev)
 	if (pcdev->pdata->flags & RCAR_VIN_FLAG_USE_8BIT_BUS)
 		flags |= SOCAM_DATAWIDTH_8;
 
+	if (pcdev->pdata->flags & RCAR_VIN_FLAG_USE_12BIT_BUS)
+		flags |= SOCAM_DATAWIDTH_12;
+
 	if (pcdev->pdata->flags & RCAR_VIN_FLAG_USE_16BIT_BUS)
 		flags |= SOCAM_DATAWIDTH_16;
 
+	if (pcdev->pdata->flags & RCAR_VIN_FLAG_USE_18BIT_BUS)
+		flags |= SOCAM_DATAWIDTH_18;
+
+	if (pcdev->pdata->flags & RCAR_VIN_FLAG_USE_24BIT_BUS)
+		flags |= SOCAM_DATAWIDTH_24;
+
 	if (flags & SOCAM_DATAWIDTH_MASK)
 		return flags;
 
@@ -982,8 +991,12 @@ static int rcar_vin_try_bus_param(struct soc_camera_device *icd,
 	camera_flags = icd->ops->query_bus_param(icd);
 	common_flags = soc_camera_bus_param_compatible(camera_flags,
 						       make_bus_param(pcdev));
-	if (!common_flags || buswidth > 16 ||
-	    (buswidth > 8 && !(common_flags & SOCAM_DATAWIDTH_16)))
+	if (!common_flags ||
+	    ((buswidth > 8) && (common_flags & SOCAM_DATAWIDTH_8)) ||
+	    ((buswidth > 12) && (common_flags & SOCAM_DATAWIDTH_12)) ||
+	    ((buswidth > 16) && (common_flags & SOCAM_DATAWIDTH_16)) ||
+	    ((buswidth > 18) && (common_flags & SOCAM_DATAWIDTH_18)) ||
+	    ((buswidth > 24) && (common_flags & SOCAM_DATAWIDTH_24)))
 		return -EINVAL;
 
 	return 0;
diff --git a/include/media/rcarvin.h b/include/media/rcarvin.h
index cce4ef7..2fb9eee 100644
--- a/include/media/rcarvin.h
+++ b/include/media/rcarvin.h
@@ -4,9 +4,12 @@
 #include <media/soc_camera.h>
 
 #define RCAR_VIN_FLAG_USE_8BIT_BUS	(1 << 0) /* use  8bit bus width */
-#define RCAR_VIN_FLAG_USE_16BIT_BUS	(1 << 1) /* use 16bit bus width */
-#define RCAR_VIN_FLAG_HSYNC_LOW		(1 << 2) /* default High if possible */
-#define RCAR_VIN_FLAG_VSYNC_LOW		(1 << 3) /* default High if possible */
+#define RCAR_VIN_FLAG_USE_12BIT_BUS	(1 << 1) /* use 12bit bus width */
+#define RCAR_VIN_FLAG_USE_16BIT_BUS	(1 << 2) /* use 16bit bus width */
+#define RCAR_VIN_FLAG_USE_18BIT_BUS	(1 << 3) /* use 18bit bus width */
+#define RCAR_VIN_FLAG_USE_24BIT_BUS	(1 << 4) /* use 24bit bus width */
+#define RCAR_VIN_FLAG_HSYNC_LOW		(1 << 5) /* default High if possible */
+#define RCAR_VIN_FLAG_VSYNC_LOW		(1 << 6) /* default High if possible */
 
 struct rcar_vin_info {
 	unsigned long flags;
-- 
1.7.0.4

